src/runtime-rs/crates/runtimes/src/manager.rs

```diff
@@ -226,6 +226,13 @@ impl RuntimeHandlerManager {
                 .await
                 .context("create container")?;
 
+            // update sandbox resource when creating container
+            // the function is idempotent
+            instance
+                .update_sandbox_resource()
```

Comment:
Hi @Ji-Xinyou, I still recommend not to do here. For my concern:

- I have not found any cgroup operations in your code, but actually we need to do that after vcpu hotpluged. Since we will create new threads, and they need to be constrained. (If you have wriiten those code, please point that to me, thanks~)
- Assumed we have to do cgroup operations, so under current implementation, for one create_container request, we have to do cgroup operations twice. And cgroup operations are always the bottlenecks in high elasticity and high density senarios.

Reply:
* In create container, the cgroup is updated.
* In start container, the cgroup does not need to be updated.
  * case1: a created but unstarted container, whose cgroup has already been updated
  * case2: a started but paused container, whose cgroup is setup upon creation
* In update container, the cgroup is updated.

In current implementation, I think the cgroup is updated only once since the sandbox resource update does not enforce the cgroup update.

Reply:
I think @YushuoEdge 's point is you need to update cgroup to constrain Vmm and vcpu threads after `update_sandbox_resource`. As cgroup is been updated in `create_container`once, so it will become twice

Reply:
You can check how kata 2.0 did this https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/sandbox.go#L1376

Reply:
> I think @YushuoEdge 's point is you need to update cgroup to constrain Vmm and vcpu threads after `update_sandbox_resource`. As cgroup is been updated in `create_container`once, it will become twice when you do it after `update_sandbox_resource`

1. The cgroup information like `cpu_quota` and `cpu_period` is set upon container creation and update(in create_container and update_container), we DO NOT need to modify this when we update the sandbox resource, all it does it to `hotplug and record` cpu/mem resource.
2. As you mentioned in 2.0, 2.0 update the resource AFTER the container creation. The reason is just like I said, this update DOES NOT modify the cgroup. However, it does update the `cpuset`, [see here](https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/sandbox.go#L2236). Since we are not supporting cpu set in the PR, this is not a problem.

Reply:
We need to write newly spawned threads to cgroup controller, right?

Reply:
> @Ji-Xinyou We need to write newly spawned threads to cgroup controller, right?

Hmm, I am not 100% sure. But I think the logic is like this.

The `spec` specifies each container's fair share of cpu, lets say container `C1` has a share of 2 vCPUs and `C2` has a share of 1 vCPUs. Their cgroup information are **contraint in the VM**, i.e., inside the VM, they are sharing the vCPU with the ratio of `2:1`.

What we really care is on the host side, which is the actual share of the container. Take QEMU as an example, each vCPU is actually a thread. In Linux CFS, the thread with the same parameters (e.g. nice value) are sharing the harts (hardware threads) **fairly**.

Inside the container, they are sharing the vCPU correctly, which means they are sharing the vCPU threads correctly, which further means they are sharing the host-side harts correctly.

Therefore, my current opinion would be `if we are not supporting cpuset, the newly spawned threads will be fine if they are correctly handled by the guest cgroup (which I think is the case).`

Reply:
Refer to https://github.com/kata-containers/kata-containers/blob/main/docs/design/host-cgroups.md
I mean, taking `sandbox_cgroup_only=false` and `cgroup v1` as an example, in the host, for each pod, we will have two cgroup_paths, /kubepods/kata_<sandbox_id> and /kata_overhead/<sandbox_id>. We name them as cgroup_controller and overhead_controller in runtime's code. We have to write the tid to each path's `tasks` so that the resources[on the host] each thread can use can be constrained.

When we have hotpluged new vcpus, there have spawned new vcpu threads. But the newly threads did not appear in any cgroup controller now, which indicates that the newly spawned vcpu threads can use any threads they want on the host.

As you have mentioned, although we do not have cpuset support now, but I think we should implement a more general  methods, which can more suitable for all host resources' management. Otherwise, we still need to refactor the code logic.

Reply:
In addition, under your current implementation, we are experiencing:
1. start vm with default resources
2. update the host resources that pod can use when create container, sometimes we may need to set contrain in the guest, according to container's spec
3. update the vm's resources

There exists a logic problem, when we set resources' limits for each container in the guest, we're setting cgroups based on resources that there should have existed. (As from upper layer's perspective, we are setting cgroups according the host's resources.) But there did not exist enough in reality. This will lead to error.  Still, taking cpuset as an example,
If we already start the vm with default 1 vcpus, then there comes a container with 2 vcpus requested. If we have to set cpuset in the guest, assumping we are going to set `0-1` in the guest, then the guest OS will report error.

Same cases will also happen when containers request for hugepage, etc.

So, what I mean here is, we should finish preparing resources before we set cgroup, which is more natural logic.